tmol/pose/pose_kinematics.py

```diff
@@ -0,0 +1,453 @@
+        nz_real_conn_block_ind_prelim,
+        nz_real_conn_conn_ind_prelim,
+    ) = torch.nonzero(real_conn, as_tuple=True)
+
+    other_block_prelim = pose_stack.inter_residue_connections[real_conn][:, 0]
+    other_conn_prelim = pose_stack.inter_residue_connections[real_conn][:, 1]
+    complete_conn = other_block_prelim != -1
+    other_block = other_block_prelim[complete_conn].to(torch.int64)
+    other_conn = other_conn_prelim[complete_conn].to(torch.int64)
+
+    nz_real_conn_pose_ind = nz_real_conn_pose_ind_prelim[complete_conn]
+    nz_real_conn_block_ind = nz_real_conn_block_ind_prelim[complete_conn]
+    nz_real_conn_conn_ind = nz_real_conn_conn_ind_prelim[complete_conn]
+
+    other_block_atom = pbt.conn_atom[
+        pose_stack.block_type_ind64[nz_real_conn_pose_ind, other_block], other_conn
+    ]
+
+    global_at1 = (
+        pose_stack.max_n_pose_atoms * nz_real_conn_pose_ind
+        + pose_stack.block_coord_offset[nz_real_conn_pose_ind, nz_real_conn_block_ind]
+        + pbt.conn_atom[
+            pose_stack.block_type_ind64[nz_real_conn_pose_ind, nz_real_conn_block_ind],
+            nz_real_conn_conn_ind,
+        ]
+    )
+    global_at2 = (
+        pose_stack.max_n_pose_atoms * nz_real_conn_pose_ind
+        + pose_stack.block_coord_offset[nz_real_conn_pose_ind, other_block]
+        + other_block_atom
+    )
+
+    interres_bonds = torch.stack((global_at1, global_at2), dim=1)
+
+    intrares_bonds = get_all_intrablock_bonds(pose_stack)
+    bonds = torch.cat((intrares_bonds, interres_bonds), dim=0)
+
+    return bonds
+
+
+def get_root_atom_indices(
+    pose_stack: PoseStack, fold_tree_roots: NDArray[int][:]
+) -> Tensor[torch.int32][:]:
+    # take the first atom from each root residue
+    assert pose_stack.n_poses == fold_tree_roots.shape[0]
+    roots64_dev = torch.tensor(
+        fold_tree_roots, dtype=torch.int64, device=pose_stack.device
+    )
+    n_pose_arange = torch.arange(
+        pose_stack.n_poses, dtype=torch.int32, device=pose_stack.device
+    )
+    return (
+        n_pose_arange * pose_stack.max_n_pose_atoms
+        + pose_stack.block_coord_offset[n_pose_arange.to(torch.int64), roots64_dev]
+    )
+
+
+def construct_pose_stack_kinforest(
+    pose_stack: PoseStack, fold_forest: FoldForest
+) -> KinForest:
+
+    intra_block_bonds = get_all_intrablock_bonds(pose_stack)
+    kin_polymeric_connections = mark_polymeric_bonds_in_foldforest_edges(
+        pose_stack.n_poses, pose_stack.max_n_blocks, fold_forest.edges
+    )
+    kin_polymeric_bonds = get_polymeric_bonds_in_fold_forest(
+        pose_stack, kin_polymeric_connections
+    )
+
+    # TO DO: add bonds between jump atoms
+    # TO DO: determine which atoms on a block a jump should
+    # connect to. Logic in R3: take the central "mainchain" atom
+    # which is only ok for polymers, but perverse for anything else.
+    # What's the mainchain of a ligand?!
+    # jump_atom_pairs = get_jump_bonds_in_fold_forest(pose_stack, fold_forest)
```

Comment:
maybe atom with minimal max #bonds to any other atom?

Reply:
I'm going to come back to this one; yes, we need the ability to create jumps, and that isn't a low priority, but it's definitely not the top of the list.